compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala

```diff
@@ -28,138 +28,221 @@ class ElimRepeated extends MiniPhase with InfoTransformer { thisPhase =>
       tp.derivedLambdaType(paramNames, paramTypes1, resultType1)
     case tp: PolyType =>
       tp.derivedLambdaType(tp.paramNames, tp.paramInfos, elimRepeated(tp.resultType))
     case tp =>
       tp
-  }
 
   def transformTypeOfTree(tree: Tree)(using Context): Tree =
     tree.withType(elimRepeated(tree.tpe))
 
+  override def transformTypeApply(tree: TypeApply)(using Context): Tree =
+    transformTypeOfTree(tree)
+
   override def transformIdent(tree: Ident)(using Context): Tree =
     transformTypeOfTree(tree)
 
   override def transformSelect(tree: Select)(using Context): Tree =
     transformTypeOfTree(tree)
 
-  override def transformApply(tree: Apply)(using Context): Tree = {
+  override def transformApply(tree: Apply)(using Context): Tree =
     val args = tree.args.mapConserve {
       case arg: Typed if isWildcardStarArg(arg) =>
         val isJavaDefined = tree.fun.symbol.is(JavaDefined)
         val tpe = arg.expr.tpe
-        if (isJavaDefined && tpe.derivesFrom(defn.SeqClass))
+        if isJavaDefined && tpe.derivesFrom(defn.SeqClass) then
           seqToArray(arg.expr)
-        else if (!isJavaDefined && tpe.derivesFrom(defn.ArrayClass))
+        else if !isJavaDefined && tpe.derivesFrom(defn.ArrayClass)
           arrayToSeq(arg.expr)
         else
           arg.expr
       case arg => arg
     }
     transformTypeOfTree(cpy.Apply(tree)(tree.fun, args))
-  }
 
   /** Convert sequence argument to Java array */
-  private def seqToArray(tree: Tree)(using Context): Tree = tree match {
+  private def seqToArray(tree: Tree)(using Context): Tree = tree match
     case SeqLiteral(elems, elemtpt) =>
       JavaSeqLiteral(elems, elemtpt)
     case _ =>
       val elemType = tree.tpe.elemType
       var elemClass = erasure(elemType).classSymbol
-      if (defn.NotRuntimeClasses.contains(elemClass)) elemClass = defn.ObjectClass
+      if defn.NotRuntimeClasses.contains(elemClass) then
+        elemClass = defn.ObjectClass
+      end if
       ref(defn.DottyArraysModule)
         .select(nme.seqToArray)
         .appliedToType(elemType)
         .appliedTo(tree, clsOf(elemClass.typeRef))
-  }
 
   /** Convert Java array argument to Scala Seq */
   private def arrayToSeq(tree: Tree)(using Context): Tree =
     tpd.wrapArray(tree, tree.tpe.elemType)
 
-  override def transformTypeApply(tree: TypeApply)(using Context): Tree =
-    transformTypeOfTree(tree)
-
-  /** If method overrides a Java varargs method, add a varargs bridge.
-   *  Also transform trees inside method annotation
+  /** If method overrides a Java varargs method or is annotated with @varargs, add a varargs bridge.
+   *  Also transform trees inside method annotation.
    */
   override def transformDefDef(tree: DefDef)(using Context): Tree =
-    atPhase(thisPhase) {
-      if (tree.symbol.info.isVarArgsMethod && overridesJava(tree.symbol))
-        addVarArgsBridge(tree)
+    val sym = tree.symbol
+    val hasAnnotation = hasVarargsAnnotation(sym)
```

Comment:
This val can be moved closer to where it's used (or even inlined since it's only used once and its definition is pretty short)

Reply:
`hasAnnotation` is also used in the `else` to emit an error

Reply:
Ah right, I missed that!